File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/Syms/PurposeReferencePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.Syms
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;
    using System;
    using System.Collections.Generic;
    using System.Text;

    class PurposeReferencePersistence
    {
        public static CdmPurposeReference FromData(CdmCorpusContext ctx, JToken obj)
        {
            if (obj == null)
            {
                return null;
            }

            bool simpleReference = true;
            bool? optional = null;
            dynamic purpose;

            if (obj is JValue)
                purpose = obj;
            else
            {
                simpleReference = false;

                if (obj["optional"] != null)
                {
                    if (bool.TryParse(obj["optional"].ToString(), out bool optVal))
                    {
                        optional = optVal;
                    }
                }

                if (obj["purposeReference"] is JValue)
                    purpose = (string)obj["purposeReference"];
                else
                    purpose = PurposePersistence.FromData(ctx, obj["purposeReference"]);
            }

            CdmPurposeReference purposeReference = ctx.Corpus.MakeRef<CdmPurposeReference>(CdmObjectType.PurposeRef, purpose, simpleReference);
            
            if (optional != null)
            {
                purposeReference.Optional = optional;
            }

            if (!(obj is JValue))
            {
                Utils.AddListToCdmCollection(purposeReference.AppliedTraits, Utils.CreateTraitReferenceList(ctx, obj["appliedTraits"]));
            }

            return purposeReference;
        }

        public static dynamic ToData(CdmPurposeReference instance, ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectRefPersistence.ToData(instance, resOpt, options);
        }
    }
}

File: objectModel/Python/cdm/persistence/syms/purpose_reference_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Union, Optional, List

from cdm.enums import CdmObjectType
from cdm.objectmodel import CdmCorpusContext, CdmPurposeReference

from . import utils
from .cdm_object_ref_persistence import CdmObjectRefPersistence
from .purpose_persistence import PurposePersistence
from .types import PurposeReference


class PurposeReferencePersistence(CdmObjectRefPersistence):
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: Union[str, PurposeReference]) -> Optional[CdmPurposeReference]:
        if not data:
            return None

        simple_reference = True
        optional = None  # type: Optional[bool]
        applied_traits = None
        purpose = None

        if isinstance(data, str):
            purpose = data
        else:
            simple_reference = False
            optional = data.optional

            if isinstance(data.purposeReference, str):
                purpose = data.purposeReference
            else:
                purpose = PurposePersistence.from_data(ctx, data.purposeReference)

        purpose_reference = ctx.corpus.make_ref(CdmObjectType.PURPOSE_REF, purpose, simple_reference)

        if optional is not None:
            purpose_reference.optional = optional

        if not isinstance(data, str):
            utils.add_list_to_cdm_collection(purpose_reference.applied_traits,
                                             utils.create_trait_reference_array(ctx, data.get('appliedTraits')))

        return purpose_reference
